File: p1.py
# Time Complexity : O(n*lgn)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this :


# Your code here along with comments explaining your approach


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:

        d = dict()

        for n in nums:
            if n in d:
                d[n] += 1
            else:
                d[n] = 1

        l = list(d.keys())
        l.sort()

        arr = [[l[i], d[l[i]]] for i in range(len(l))]

        dp = [[0, 0] for i in range(len(arr))]

        dp[0][0] = arr[0][1] * arr[0][0]
        dp[0][1] = 0

        for i in range(1, len(arr)):

            if arr[i][0] == arr[i - 1][0] + 1:
                dp[i][0] = dp[i - 1][1] + arr[i][1] * arr[i][0]
            else:
                dp[i][0] = max(dp[i - 1][0], dp[i - 1][1]) + arr[i][1] * arr[i][0]

            dp[i][1] = max(dp[i - 1][0], dp[i - 1][1])

        return max(dp[-1])

File: p1.java
// Time Complexity : time complexity would be O(n) + O(max(n))
// Space Complexity : O(n)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach

class Solution {
    public int deleteAndEarn(int[] nums) {
        //Using hashmap to store values and their frequency
        Map<Integer, Integer> map = new HashMap<>();
        //Variable for getting min and max value
        int min = 20001;
        int max = 0;
        // Iterating through the array to get the min and max value and storing values and frequency in hashmap
        for(int i=0; i< nums.length; i++){
            min = Math.min(min, nums[i]);
            max = Math.max(max, nums[i]);
            map.put(nums[i], map.getOrDefault(nums[i],0)+1);
        }

        
        
        // Getting value for previous and current pointer from the hashmap
        int prev = map.get(min)*min;
        int curr = Math.max(prev, map.getOrDefault(min+1,0)*(min+1));
        System.out.println(prev);
        System.out.println(curr);
        //Iterating through the hashmap to get the max number of points
        for(int i = min+2; i<=max; i++){
            int temp = curr;
            curr = Math.max(curr, (map.getOrDefault(i,0)*(i))+prev);
            prev = temp;
        }

        return curr;
    }
}